### PR TITLE
Service page: Sidebar Labels filter

### PR DESCRIPTION
**Pull request by pierlo-upitup**
This adds the sidebar labels filter to the Services Page.

labels12

Verified by @leemunroe ™️ 

NOTES
----

I am having trouble getting the dropdown ellipsis styling to look right

- [x] when using `.text-overflow` inside a dropdown element, a weird hover artifact can be seen:
image

- [x] Integration tests pass locally
image

- [x] how can we avoid having the persistent item (`Labels`) to show up inside of the open dropdown?

OUT OF SCOPE
----

- [ ] at some point we'll have to introduce a hash map so we can avoid storing the full key/value label strings in the URL – with really long labels this may end up getting us into trouble – out of scope

### CODE REVIEW

**Comment by mesosphere-web** [comment]
Reviewer: @aldipower

**Comment by jfurrow** [comment]
@pierlo-upitup I hope you don't mind but I pushed a small commit that should fix the `.text-overflow` oddities and hides the persistent ID from the dropdown menu. See the commit for how I did it: https://github.com/dcos/dcos-ui/pull/433/commits/ec9c289e6e550fd56d452dd88e9a021335236503

**Comment by pierlo-upitup** [comment]
Thank you @jfurrow !!!

File: src/js/components/SidebarLabelsFilter.js [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+    const labelClassName = classNames(
+      'label-pill badge text-muted text-overflow inverse'
+    );
+
+    const removeLabelClassNames = classNames(
+      'badge text-muted inverse clickable text-align-center remove-filter'
+    );
+
+    let labelNodes = selectedLabels.map(({key, value}, i) => {
```
Comment (aldipower) [inline src/js/components/SidebarLabelsFilter.js]: I like this notation of `{key, value}` somehow. :)

File: src/js/components/SidebarLabelsFilter.js [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+      let labelText = `${key} : ${value}`;
+      return (
+        <li className={labelClassName} key={i} title={labelText}>
+          <div className="text-overflow pill-wrap">
+            <span className="text-overflow">{labelText}</span>
+            <a className={removeLabelClassNames}
+               onClick={this.handleActionSelection.bind(this, {key, value})}>
```
Comment (aldipower) [inline src/js/components/SidebarLabelsFilter.js]: Thanx for indenting this way. :)

File: src/js/components/SidebarLabelsFilter.js [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+      <ul className={labelsListClassNames}>
+        {labelNodes}
+      </ul>
+    )
+  }
+
+  handleActionSelection({key, value}) {
+    const {selectedLabels} = this.state;
+    const stringify = JSON.stringify;
+    let nextSelectedLabels = selectedLabels.slice();
```
Comment (aldipower) [inline src/js/components/SidebarLabelsFilter.js]: Well distinguished between `const` and `let` here! AAA

File: src/js/components/SidebarLabelsFilter.js [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+    let labelIndex = selectedLabels.findIndex(function (item) {
```
Comment (aldipower) [inline src/js/components/SidebarLabelsFilter.js]: Ah, the `labelIndex` could be a `const` too. :)
But I don't care, only change if you like.

File: src/js/structs/ServiceTree.js [hunk 5]
```diff
@@ -120,6 +121,21 @@ module.exports = class ServiceTree extends Tree {
...
       }
 
+      if (filter.labels != null && filter.labels.length > 0) {
+        services = services.filter(function (service) {
+          return filter.labels.some(function (label) {
+            let serviceLabels = ServiceUtil.convertServiceLabelsToArray(
+              service
+            );
+
+            return serviceLabels.find(function (serviceLabel) {
```
Comment (aldipower) [inline src/js/structs/ServiceTree.js]: This `.find` could be replaced by `.some` to get rid of the ` != null` at the end.

File: tests/pages/services/SidebarFilter-cy.js [hunk 6]
```diff
@@ -66,6 +74,8 @@ describe('Sidebar Filter', function () {
       cy.get('tbody tr:visible').should('to.have.length', 1);
     });
 
+
```
Comment (aldipower) [inline tests/pages/services/SidebarFilter-cy.js]: If you need more space buy a house at the countryside.. ;)

**Comment by aldipower** [comment]
Code looks too good to me. Please add some PHP.
Very good and understandable work.
Now I am doing a functionality test.

**Comment by aldipower** [comment]
Filtering, selecting and so on works perfectly!
But there is a displaying problem of the dropdown-arrow in Firefox, Chrome works.
dropwdownarrow

**Comment by aldipower** [comment]
Done with review iteration 1.

**Comment by pierlo-upitup** [comment]
And this is how it looks in Safari:

image

As I followed @MatApple's suggestion on the dropdown styling, I suspect we have the same problem everywhere we use the `.dropdown-wide` class at the minute. My suggestion is to address this in a separate PR.

**Comment by aldipower** [comment]
OK, fine with addressing this separately.

**Comment by pierlo-upitup** [comment]
Thanks for the review @aldipower – I also spotted a bug while refactoring the filtering code part and here's proof that everything works as it should:

image

**Comment by aldipower** [comment]
Geil, Alter!

**Comment by pierlo-upitup** [comment]
@mesosphere-ci retest this please

**Comment by Poltergeist** [comment]
It also works with unicode 👍 
image
It haz tests!
image
Which pass local
image
image

**Pull request merged**